precompile/reward_manager.go

```diff
@@ -0,0 +1,457 @@
+)
+
+// Singleton StatefulPrecompiledContract and signatures.
+var (
+	_ StatefulPrecompileConfig = &RewardManagerConfig{}
+
+	ErrCannotAllowFeeRecipients      = errors.New("non-enabled cannot call allowFeeRecipients")
+	ErrCannotAreFeeRecipientsAllowed = errors.New("non-enabled cannot call areFeeRecipientsAllowed")
+	ErrCannotCurrentRewardAddress    = errors.New("non-enabled cannot call currentRewardAddress")
+	ErrCannotDisableRewards          = errors.New("non-enabled cannot call disableRewards")
+	ErrCannotSetRewardAddress        = errors.New("non-enabled cannot call setRewardAddress")
+
+	ErrCannotEnableBothRewards = errors.New("cannot enable both fee recipients and reward address at the same time")
+	ErrEmptyRewardAddress      = errors.New("reward address cannot be empty")
+
+	RewardManagerABI        abi.ABI                     // will be initialized by init function
+	RewardManagerPrecompile StatefulPrecompiledContract // will be initialized by init function
+
+	rewardAddressStorageKey        = common.Hash{'r', 'a', 's', 'k'}
+	allowFeeRecipientsAddressValue = common.Hash{'a', 'f', 'r', 'a', 'v'}
+)
+
+type InitialRewardConfig struct {
+	AllowFeeRecipients bool           `json:"allowFeeRecipients"`
+	RewardAddress      common.Address `json:"rewardAddress,omitempty"`
+}
+
+func (i *InitialRewardConfig) Verify() error {
+	switch {
+	case i.AllowFeeRecipients && i.RewardAddress != (common.Address{}):
+		return ErrCannotEnableBothRewards
+		// shall we also check blackhole address here?
+	default:
+		return nil
+	}
+}
+
+func (c *InitialRewardConfig) Equal(other *InitialRewardConfig) bool {
+	if other == nil {
+		return false
+	}
+
+	return c.AllowFeeRecipients == other.AllowFeeRecipients && c.RewardAddress == other.RewardAddress
+}
+
+func (i *InitialRewardConfig) Configure(state StateDB) {
+	// enable allow fee recipients
+	if i.AllowFeeRecipients {
+		EnableAllowFeeRecipients(state)
+	} else if i.RewardAddress == (common.Address{}) {
+		// if reward address is empty and allow fee recipients is false
+		// then disable rewards
+		DisableFeeRewards(state)
```

Comment:
Since `DisableFeeRewards` is identical to setting the reward address to the burn address, would it make sense to get rid of `DisableFeeRewards` helper and just store the specified reward address with no further check? I don't think it should be necessary to re-route the zero address to the burn address that we have used in the past.

We could also get rid of the `disableRewards` function since it does the same thing as `setRewardAddress` with a special case as well.

Reply:
In the code-wise you're absolutely correct, just wanted to improve UX by adding this function. I don't think it's very practical to call the function with blackhole address each time, especially from external contracts. We don't provide burn address in EVM to contracts, so I think it's very likely to have a mistake in those contracts. Like having a wrong blackhole address etc.